File: d07/d07.exs
hands = File.stream!("input.txt")
|> Stream.map(fn line -> 
  [hand, bid] = line |> String.trim() |> String.split(" ")
  {to_charlist(hand), String.to_integer(bid)}
end)

hand_strength = [[5], [4], [3, 2], [3], [2, 2], [2], [1], []]
card_strength = %{false: '23456789TJQKA', true: 'J23456789TQKA'}

order = fn hand, use_jokers? ->
  {rest, jokers} = use_jokers? && Enum.split_with(hand, & &1 != ?J) || {hand, []}
  cnts = rest |> Enum.frequencies() |> Map.values() |> Enum.frequencies()

  hand_indices = hand
  |> Enum.map(fn card -> 
    Enum.find_index(card_strength[use_jokers?], & &1 == card) 
  end)

  idx = hand_strength
  |> Enum.find_index(fn need -> 
    need
    |> Enum.frequencies()
    |> Enum.all?(fn {k, v} -> v <= (cnts[k] || 0) end) 
  end)

  [head | rest] = hand_strength
  |> Enum.at(idx)
  |> then(& length(&1)>0 && &1 || [0])
  
  extra = use_jokers? && length(jokers) || 0
  strength = hand_strength
  |> Enum.find_index(& &1 == [head+extra | rest])

  {strength * -1, hand_indices}
end

solve = fn use_jokers? ->
  hands
  |> Enum.sort_by(fn {hand, _bid} -> order.(hand, use_jokers?) end)
  |> Enum.with_index(1)
  |> Enum.map(fn {{_hand, bid}, i} -> bid * i end)
  |> Enum.sum()
end

IO.inspect(solve.(false), label: "Star 1")
IO.inspect(solve.(true), label: "Star 2")

File: d07/d07.py
from collections import Counter

l = [(a, int(b)) for a, b in map(str.split, open(0).readlines())]
hand_pow = [[5], [4], [3, 2], [3], [2, 2], [2], [1], []]
card_pow = ['23456789TJQKA', 'J23456789TQKA']

def order(hand, use_js):
    *rest, = filter(lambda c: not use_js or c != 'J', hand)
    num_js = use_js and sum(c == 'J' for c in hand) or 0
    head, *tail = next(filter(lambda need: Counter(need) <= Counter(Counter(rest).values()), hand_pow)) or [0]
    return (-hand_pow.index([head+num_js, *tail]), tuple(map(card_pow[use_js].index, hand)))

print([sum(i*bid for i, (_, bid) in enumerate(sorted(l, key=lambda x: order(x[0], use_js)), 1)) for use_js in [0, 1]])
